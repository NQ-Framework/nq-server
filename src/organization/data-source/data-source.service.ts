import { Injectable } from '@nestjs/common';
import { getFirebaseApp } from '../../firebase/initialize';
import { LoggerService } from '../../logger/logger.service';
import { OrganizationMember, DataSource } from '@nqframework/models';

@Injectable()
export class DataSourceService {
  constructor(private logger: LoggerService) {}

  async getDataSourceConfigurations(
    organizationId: string,
  ): Promise<{ members: OrganizationMember[]; dataSources: DataSource[] }> {
    this.logger.debug(
      `Getting Data Source Configuration for ${organizationId}`,
    );
    const app = await getFirebaseApp();
    const document = await app
      .firestore()
      .doc(`organizations/${organizationId}`)
      .get();
    if (!document.exists) {
      this.logger.warn(
        `Looking up configuration for non existant organization: ${organizationId}`,
      );
      throw new Error('Invalid organization id');
    }
    const data = document.data() as FirebaseFirestore.DocumentData;
    return { members: data.members, dataSources: data.dataSources };
  }
}
