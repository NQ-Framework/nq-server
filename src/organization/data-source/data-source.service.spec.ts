import { Test, TestingModule } from '@nestjs/testing';
import { LoggerModule } from '../../logger/logger.module';
import { DataSourceService } from './data-source.service';
const mockMembers = [1, 2, 3];
const mockDataSources = [4, 5, 6];
jest.mock('../../firebase/initialize', () => ({
  getFirebaseApp: () => ({
    firestore: () => ({
      doc: (organizationId: string) => ({
        get: () => {
          if (organizationId.includes('non-existent')) {
            return { exists: false };
          }
          return {
            exists: true,
            data: () => ({
              members: mockMembers,
              dataSources: mockDataSources,
            }),
          };
        },
      }),
    }),
  }),
}));

describe('DataSourceService', () => {
  let service: DataSourceService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [LoggerModule],
      providers: [DataSourceService],
    }).compile();

    service = module.get<DataSourceService>(DataSourceService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
  it('should throw for non existent organization id', async () => {
    await expect(
      service.getDataSourceConfigurations('test non-existent organization id'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should return organization configuration data', async () => {
    await expect(
      service.getDataSourceConfigurations('test organization id'),
    ).resolves.toEqual({
      members: mockMembers,
      dataSources: mockDataSources,
    });
  });
});
