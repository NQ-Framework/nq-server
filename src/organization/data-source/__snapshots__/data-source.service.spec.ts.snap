// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSourceService should throw for non existent organization id 1`] = `"Invalid organization id"`;
